Chaincase.Common/Services/AesThenHmac.cs

```diff
@@ -0,0 +1,402 @@
+        /// <remarks>
+        /// Significantly less secure than using random binary keys.
+        /// </remarks>
+        public static string DecryptWithPassword(string encryptedMessage, string password,
+                                 int nonSecretPayloadLength = 0)
+        {
+            if (string.IsNullOrWhiteSpace(encryptedMessage))
+                throw new ArgumentException("Encrypted Message Required!", "encryptedMessage");
+
+            var cipherText = Convert.FromBase64String(encryptedMessage);
+            var plainText = DecryptWithPassword(cipherText, password, nonSecretPayloadLength);
+            return plainText == null ? null : Encoding.UTF8.GetString(plainText);
+        }
+
+        public static byte[] Encrypt(byte[] secretMessage, byte[] cryptKey, byte[] authKey, byte[] nonSecretPayload = null)
+        {
+            //User Error Checks
+            if (cryptKey == null || cryptKey.Length != KeyBitSize / 8)
+                throw new ArgumentException(String.Format("Key needs to be {0} bit!", KeyBitSize), "cryptKey");
+
+            if (authKey == null || authKey.Length != KeyBitSize / 8)
+                throw new ArgumentException(String.Format("Key needs to be {0} bit!", KeyBitSize), "authKey");
+
+            if (secretMessage == null || secretMessage.Length < 1)
+                throw new ArgumentException("Secret Message Required!", "secretMessage");
+
+            //non-secret payload optional
+            nonSecretPayload = nonSecretPayload ?? new byte[] { };
+
+            byte[] cipherText;
+            byte[] iv;
+
+            using (var aes = new AesManaged
+            {
+                KeySize = KeyBitSize,
+                BlockSize = BlockBitSize,
+                Mode = CipherMode.CBC,
+                Padding = PaddingMode.PKCS7
+            })
+            {
+
+                //Use random IV
+                aes.GenerateIV();
+                iv = aes.IV;
+
+                using (var encrypter = aes.CreateEncryptor(cryptKey, iv))
+                using (var cipherStream = new MemoryStream())
+                {
+                    using (var cryptoStream = new CryptoStream(cipherStream, encrypter, CryptoStreamMode.Write))
+                    using (var binaryWriter = new BinaryWriter(cryptoStream))
+                    {
+                        //Encrypt Data
+                        binaryWriter.Write(secretMessage);
+                    }
+
+                    cipherText = cipherStream.ToArray();
+                }
+
+            }
+
+            //Assemble encrypted message and add authentication
+            using (var hmac = new HMACSHA256(authKey))
+            using (var encryptedStream = new MemoryStream())
+            {
+                using (var binaryWriter = new BinaryWriter(encryptedStream))
+                {
+                    //Prepend non-secret payload if any
```

Comment:
what's the purpose of the 'non-secret' payload?

Reply:
aka "associated data"

e.g. the salt here:

https://github.com/chaincase-app/Chaincase/blob/71e2725ac8e28b07154fbb254546ae0abc662238/Chaincase.Common/Services/AesThenHmac.cs#L336-L369